### PR TITLE
Move params mangling to controller

### PR DESCRIPTION
**Pull request by mbj**
Extraction from #264

### CODE REVIEW

**Comment by mbj** [comment]
@dkubb This is also ready for review.

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 1]
```diff
@@ -93,6 +91,63 @@ def rescue_from_spree_gateway_error(exception)
...
+    def checkout_attributes
+      params.fetch(
+        :order,
+        ActionController::Parameters.new
+      ).permit(PERMITTED_CHECKOUT_ATTRIBUTES)
+    end
+
+    # The payment source attributes
+    #
+    # @return [ActionController::Parameters, nil]
```
Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: This `nil` is invalid and should be removed.

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -87,15 +87,12 @@ steps:
...
+  - name: ':skull: Spree::CheckoutController'
     command: |
       docker-wrapper --path frontend --database -- bash -c '
         ../scripts/setup-ci.sh frontend
         bundle exec mutant --use rspec --require ./spec/dummy/config/environment --since HEAD~1 -- \
-         Spree::CheckoutController#post_update_redirect_path                                       \
-         Spree::CheckoutController#render_unprocessable_entity                                     \
-         Spree::CheckoutController#rescue_from_spree_gateway_error                                 \
-         Spree::CheckoutController#update
+         Spree::CheckoutController
```
Comment (dkubb) [inline .buildkite/pipeline.yml]: Should this be `Spree::CheckoutController*` for consistency with the other full mutant steps in this config?

I realize there are probably nothing besides this class, but it's good for consistency.

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 3]
```diff
@@ -42,9 +42,7 @@ def edit
 
     # Updates the order and advances to the next state (when possible.)
     def update
-      unless @order.update_from_params(params, PERMITTED_CHECKOUT_ATTRIBUTES, env)
-        return render_unprocessable_entity
-      end
+      return render_unprocessable_entity unless update_from_params
```
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: Can this be moved into a `before_action`, eg:

```ruby
before_action :render_unprocessable_entity, only: :update, unless: :update_from_params
```

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 4]
```diff
@@ -93,6 +91,63 @@ def rescue_from_spree_gateway_error(exception)
...
+    # @api private
+    def update_from_params
+      @order.update_attributes(checkout_attributes.merge(payments_attributes))
+    end
+
+    # Effective payment attributes
+    #
+    # @return [ActionController::Parameters]
+    def payments_attributes
+      attributes = checkout_attributes[:payments_attributes].try(:first)
```
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: What about `&.first` instead of `try(:first)`?

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 5]
```diff
@@ -93,6 +91,63 @@ def rescue_from_spree_gateway_error(exception)
...
+
+      return {} unless attributes
+
+      {
+        payments_attributes: [
+          attributes.merge(
+            request_env: env,
+            amount:      @order.total
+          ).merge(payment_source_attributes)
```
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: I wonder if you could write this as:

```ruby
attributes.merge(
  request_env: env,
  amount:      @order.total,
  **payment_source_attributes
)
```

File: frontend/spec/controllers/spree/checkout_controller_spec.rb [hunk 6]
```diff
@@ -1,14 +1,20 @@
...
   describe '#update' do
-    let(:post_params)    { { state: 'address' } }
-    let(:update_success) { true                 }
+    let(:post_params)    { { state: 'address' }  }
+    let(:update_success) { true                  }
+    let(:env)            { instance_double(Hash) }
 
     def do_request
-      spree_post :update, post_params.dup
+      spree_post :update, post_params.deep_stringify_keys
```
Comment (dkubb) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: I like this change. It gets you the `dup` behaviour, but it also allows you to have nicer inline setup within the specs. I wonder if it does the equivalent of a "deep dup"?

File: api/config/routes.rb [hunk 7]
```diff
@@ -13,7 +13,7 @@
 
     resources :orders, only: [] do
       resources :line_items, only: %i[destroy update]
-      resources :payments do
+      resources :payments, only: %i[show update] do
```
Comment (dkubb) [inline api/config/routes.rb]: OT but I wish `only` was a required attribute. I wish we had to specify it in all `resources` usage.

Comment (dkubb) [inline api/config/routes.rb]: Not a blocker, but in theory this change is not directly related to the subject of this commit, so it could be broken off into it's own commit in isolation.

File: frontend/app/controllers/spree/orders_controller.rb [hunk 8]
```diff
@@ -72,7 +72,7 @@ def empty
 
       def order_params
         if params[:order]
-          params[:order].permit(*permitted_order_attributes)
+          params[:order].permit(line_items_attributes: PermittedAttributes.line_item_attributes)
```
Comment (dkubb) [inline frontend/app/controllers/spree/orders_controller.rb]: I wonder if this whole block could be simplified to:

```ruby
def order_params
  params.fetch(:order, ActionController::Parameters.new)
    .permit(line_items_attributes: PermittedAttributes.line_item_attributes)
end
```

This also forces the method to return objects of the same type, rather than sometimes an `ActionController::Parameters` instance and sometimes a `Hash`.

File: .buildkite/pipeline.yml (hunk 2 above)
Comment (mbj) [inline .buildkite/pipeline.yml]: I agree to that.

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 4 above)
Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: This would raise an exception in case an object is submitted that does not respond to `#first`. AFAIK the original code we are moving used `#try` as kind of an implicit weak type check.

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 5 above)
Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: thats better, but we have to test if `ActionController::Parameters` likes to be splatted like this.

File: frontend/spec/controllers/spree/checkout_controller_spec.rb (hunk 6 above)
Comment (mbj) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: > I wonder if it does the equivalent of a "deep dup"?

AFAIK only the hash itself, not the values.

**Comment by mbj** [comment]
@dkubb This is ready for the next pass.

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 4 above)
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: Oh what objects can `checkout_attributes[:payments_attributes]` return that would not respond to `#first`?

If somehow `checkout_attributes[:payments_attributes]` returns `'1'`, then wouldn't we want to explode instead of silently ignore it?

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 3 above)
Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: @dkubb It can, but I do not see a reason to do it. Mostly because `#update_from_params` is the "main" thing (plus some other weird state mutations mandated by the to be cleaned up spree API) this action does.

Moving it into a before hook as its not a real pre-condition work feels weird. Also the conditional would be removed from mutants scope.

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 4 above)
Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: We do not like having users being able to cause explosions in the `5xx` range from bad input, it should be in the `4xx` range.

This change only intends to move the params mangling to the checkout self monkeypatch to the controller, on future passes, I'd like to apply a real schema allowing correct mapping to `4xx` on violations.

`ActionController::Parameters` is very weak and does not differentiate between types in its input.

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 9]
```diff
@@ -93,6 +91,64 @@ def rescue_from_spree_gateway_error(exception)
...
+
+    # The payment source attributes
+    #
+    # @return [ActionController::Parameters]
+    #
+    # @api private
+    def payment_source_attributes
+      payment_source = params[:payment_source]
+
+      return ActionController::Parameters.new unless payment_source
```
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: Should `ActionController::Parameters.new` be a frozen constant in this class and replaced here and the method above?

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 3 above)
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: Oh I was thinking `update_from_params` is a boolean. I didn't realize it's something with a side-effect.

If it were just a boolean test then I would probably push to have this extracted into a `before_action`, but it doesn't make sense in this case.

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 4 above)
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: Ok. Since it won't be a regression to keep this as-is, I am fine with deferring any changes until we have a way to apply a schema to the inputs.

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 3 above)
Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: Yeah it mangles the params into a mass assignment. I'd prefer we'd have an explicit intermediary stage, but this is only the first step to correctly specify and place the logic we do have to deal with.

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 9 above)
Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: @dkubb Unfortunately the implementation of `ActionController::Parameters` does mutate the state on pure get like computations like `#fetch`, no comment.

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 10]
```diff
@@ -93,6 +91,64 @@ def rescue_from_spree_gateway_error(exception)
...
+        ]
+      }
+    end
+
+    # The permitted checkout attributes
+    #
+    # @return [ActionController::Parameters]
+    #
+    # @api private
+    def checkout_attributes
```
Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: Yeah, I like that one more, good catch.

**Comment by mbj** [comment]
@dkubb Ready for the next pass.
